Index: avahi-0.6.31/avahi-core/browse-service.c
===================================================================
--- avahi-0.6.31.orig/avahi-core/browse-service.c	2010-08-25 20:51:38.983153000 -0400
+++ avahi-0.6.31/avahi-core/browse-service.c	2013-12-23 12:26:59.993833324 -0500
@@ -61,7 +61,9 @@
     flags &= AVAHI_LOOKUP_RESULT_CACHED | AVAHI_LOOKUP_RESULT_MULTICAST | AVAHI_LOOKUP_RESULT_WIDE_AREA;
 
     if (record) {
-        char service[AVAHI_LABEL_MAX], type[AVAHI_DOMAIN_NAME_MAX], domain[AVAHI_DOMAIN_NAME_MAX];
+        char service[AVAHI_LABEL_MAX] = {'\0',};
+        char type[AVAHI_DOMAIN_NAME_MAX] = {'\0',};
+        char domain[AVAHI_DOMAIN_NAME_MAX] = {'\0',};
 
         assert(record->key->type == AVAHI_DNS_TYPE_PTR);
 
