Index: hostapd-20130405/src/drivers/driver_nl80211.c
===================================================================
--- hostapd-20130405.orig/src/drivers/driver_nl80211.c	2014-03-11 16:11:18.581197118 -0400
+++ hostapd-20130405/src/drivers/driver_nl80211.c	2014-03-11 16:20:19.057196699 -0400
@@ -1583,12 +1583,21 @@
 static void mlme_event_join_ibss(struct wpa_driver_nl80211_data *drv,
 				 struct nlattr *tb[])
 {
+	struct i802_bss *bss;
+	int ifidx = -1;
+
+	ifidx = nla_get_u32(tb[NL80211_ATTR_IFINDEX]);
+	for (bss = &drv->first_bss; bss; bss = bss->next) {
+		if (bss->ifindex == ifidx)
+			break;
+	}
 	if (tb[NL80211_ATTR_MAC] == NULL) {
 		wpa_printf(MSG_DEBUG, "nl80211: No address in IBSS joined "
 			   "event");
 		return;
 	}
 	os_memcpy(drv->bssid, nla_data(tb[NL80211_ATTR_MAC]), ETH_ALEN);
+	bss->freq = drv->assoc_freq;
 	drv->associated = 1;
 	wpa_printf(MSG_DEBUG, "nl80211: IBSS " MACSTR " joined",
 		   MAC2STR(drv->bssid));
@@ -7076,6 +7085,7 @@
 	if (params->fixed_freq) {
 		wpa_printf(MSG_DEBUG, "  * fixed_freq");
 		NLA_PUT_FLAG(msg, NL80211_ATTR_FREQ_FIXED);
+		drv->assoc_freq = params->freq;
 	}
 
 	if (params->beacon_interval > 0) {
