Index: hostapd-20130405/src/drivers/driver_nl80211.c
===================================================================
--- hostapd-20130405.orig/src/drivers/driver_nl80211.c	2014-03-11 18:14:17.000000000 -0400
+++ hostapd-20130405/src/drivers/driver_nl80211.c	2014-03-11 18:19:25.829191138 -0400
@@ -311,6 +311,9 @@
 				  unsigned int freq, unsigned int wait,
 				  const u8 *buf, size_t buf_len, u64 *cookie,
 				  int no_cck, int no_ack, int offchanok);
+static int nl80211_register_frame(struct i802_bss *bss,
+				  struct nl_handle *nl_handle,
+				  u16 type, const u8 *match, size_t match_len);
 static int wpa_driver_nl80211_probe_req_report(struct i802_bss *bss,
 					       int report);
 #ifdef ANDROID
@@ -1585,6 +1588,7 @@
 {
 	struct i802_bss *bss;
 	int ifidx = -1;
+	u16 type = (WLAN_FC_TYPE_MGMT << 2) | (WLAN_FC_STYPE_AUTH << 4);
 
 	ifidx = nla_get_u32(tb[NL80211_ATTR_IFINDEX]);
 	for (bss = &drv->first_bss; bss; bss = bss->next) {
@@ -1598,6 +1602,11 @@
 	}
 	os_memcpy(drv->bssid, nla_data(tb[NL80211_ATTR_MAC]), ETH_ALEN);
 	bss->freq = drv->assoc_freq;
+
+	/* register for any AUTH message */
+	nl80211_register_frame(&drv->first_bss, drv->first_bss.nl_mgmt,
+				type, NULL, 0);
+
 	drv->associated = 1;
 	wpa_printf(MSG_DEBUG, "nl80211: IBSS " MACSTR " joined",
 		   MAC2STR(drv->bssid));
