Index: controller/splash/splash.lua
===================================================================
--- controller/splash/splash.lua	(revision 8676)
+++ controller/splash/splash.lua	(working copy)
@@ -57,6 +57,8 @@
 	local ip = luci.http.getenv("REMOTE_ADDR") or "127.0.0.1"
 	local mac = luci.sys.net.ip4mac(ip:match("^[\[::ffff:]*(%d+.%d+%.%d+%.%d+)\]*$"))
 	local blacklisted = false
+  local requesturi = luci.model.uci.cursor():get_bool("luci_splash", "general", "requesturi")
+  local homepage = luci.model.uci.cursor():get("luci_splash", "general", "homepage")
 	if mac and luci.http.formvalue("accept") then
 		uci:foreach("luci_splash", "blacklist",
         	        function(s) if s.mac:lower() == mac or s.mac == mac then blacklisted = true end
@@ -65,7 +67,11 @@
 			luci.http.redirect(luci.dispatcher.build_url("splash" ,"blocked"))
 		else
 			os.execute("luci-splash lease "..mac.." >/dev/null 2>&1")
-			luci.http.redirect(luci.model.uci.cursor():get("freifunk", "community", "homepage"))
+      if requesturi or homepage == "" then
+			  luci.http.redirect(luci.http.getenv("REQUEST_URI"))
+      else
+			  luci.http.redirect(homepage)
+      end
 		end
 	else
 		luci.http.redirect(luci.dispatcher.build_url())
Index: view/splash/splash.htm
===================================================================
--- view/splash/splash.htm	(revision 8676)
+++ view/splash/splash.htm	(working copy)
@@ -14,15 +14,15 @@
 local fs = require "luci.fs"
 local has_custom_splash = fs.access("/usr/lib/luci-splash/splashtext.html")
 
+local community, homepage, leasetime, limit_up, limit_down, R
+
 function expand (e)        
 	return (string.gsub(e, "###([A-Z_]+)###", R))
 end
 
-local community, homepage, leasetime, limit_up, limit_down, R
+local contacturl = luci.dispatcher.build_url("splash", "contact")
 
-local contacturl = luci.dispatcher.build_url("freifunk", "contact")
-
-local c = luci.model.uci.cursor():get_all("freifunk", "community")
+local c = luci.model.uci.cursor():get_first("system", "system")
 if c and c.name then
 	community = c.name
 else
