Index: wpa_supplicant.sh
===================================================================
--- wpa_supplicant.sh	(revision 31639)
+++ wpa_supplicant.sh	(working copy)
@@ -4,6 +4,7 @@
 	local key="$key"
 	local options="$3"
 	local freq=""
+	local scan_ssid='scan_ssid=1'
 	[ -n "$4" ] && freq="frequency=$4"
 
 	# wpa_supplicant should use wext for mac80211 cards
@@ -60,7 +61,13 @@
 		;;
 		*psk*)
 			key_mgmt='WPA-PSK'
-			[ "$mode" = "adhoc" -a "$driver" != "nl80211" ] && key_mgmt='WPA-NONE'
+			[ "$mode" = "adhoc" ] &&  {
+				key_mgmt='WPA-NONE'
+				scan_ssid='scan_ssid=0'
+				pairwise='pairwise=NONE'
+				group='group=TKIP'
+				ap_scan='ap_scan=2'
+			}
 			config_get_bool usepassphrase "$vif" usepassphrase 1
 			if [ "$usepassphrase" = "1" ]; then
 				passphrase="psk=\"${key}\""
@@ -124,9 +131,10 @@
 	rm -rf /var/run/wpa_supplicant-$ifname
 	cat > /var/run/wpa_supplicant-$ifname.conf <<EOF
 ctrl_interface=/var/run/wpa_supplicant-$ifname
+$ap_scan
 network={
 	$modestr
-	scan_ssid=1
+	$scan_ssid
 	ssid="$ssid"
 	$bssid
 	key_mgmt=$key_mgmt
